### PR TITLE
Port to MathComp 2

### PR DESCRIPTION
**Pull request by proux01**
Requires https://github.com/jasmin-lang/coqword/pull/22

### CODE REVIEW

File: proofs/compiler/x86_instr_decl.v [hunk 1]
```diff
@@ -860,11 +860,12 @@ Definition x86_VPUNPCKL ve sz := x86_u128_binop (@wpunpckl sz ve).
 
 (* ---------------------------------------------------------------- *)
 
+(* Import mathcomp.word.word. *)
 Definition wpblendw (m : u8) (w1 w2 : word U128) :=
   let v1 := split_vec U16 w1 in
   let v2 := split_vec U16 w2 in
   let b := split_vec 1 m in
-  let r := map3 (λ (b0 : word.word_ringType 0) (v3 v4 : mathcomp.word.word.word U16), if b0 == 1%R then v4 else v3) b v1 v2 in
+  let r := map3 (λ (b0 : (mathcomp.word.word.word 1 : ringType)) (v3 v4 : mathcomp.word.word.word U16), if b0 == 1%R then v4 else v3) b v1 v2 in
```
Comment (bgregoir) [inline proofs/compiler/x86_instr_decl.v:868]: word.word_ringType 0 is the same than 
mathcomp.word.word.word 1 ?

File: proofs/lang/expr.v [hunk 2]
```diff
@@ -534,7 +527,7 @@ Variant saved_stack :=
 Definition saved_stack_beq (x y : saved_stack) :=
   match x, y with
   | SavedStackNone, SavedStackNone => true
-  | SavedStackReg v1, SavedStackReg v2 => v1 == v2
+  | (SavedStackReg v1), SavedStackReg v2 => v1 == v2
```
Comment (bgregoir) [inline proofs/lang/expr.v:530]: Why parenthesis here?

File: proofs/lang/memory_model.v [hunk 3]
```diff
@@ -722,7 +722,8 @@ Proof.
   have [hle hlt] := wunsigned_range (p - wrepr Uptr sz).
   have := Z.mod_le _ _ hle (wsize_size_pos ws).
   have := Z_mod_lt (wunsigned (p - wrepr Uptr sz)) (wsize_size ws) ltac:(done).
-  by Psatz.lia.
+  move=> [? ?] ?; split; first exact: Zle_minus_le_0.
```
Comment (bgregoir) [inline proofs/lang/memory_model.v:725]: Psatz.lia does not work anymore with the change? Or it was very slow ?

File: proofs/lang/word.v [hunk 4]
```diff
@@ -2136,6 +2136,7 @@ Proof.
 
   rewrite Z.shiftl_mul_pow2; last done.
   split; first Psatz.lia.
-  rewrite /wbase (modulusD 64 64) modulusE -expZE /=.
-  exact: Zmult_lt_compat_r.
+  rewrite /wbase (modulusD 64 64) modulusE -expZE.
```
Comment (bgregoir) [inline proofs/lang/word.v:2139]: This seem much more complicated. What is the motivation for the change ?

**Review by bgregoir — COMMENTED** [review]
Thank a lot for this PR, I have the impression that a lot of boilerplate are simplified.
I have added some small remarks.
Also what is the level of maturity of mathcomp2 ?
I notice that our code is simplified but except that did you gain other things with this change ? I notice that we need a much recent version of Coq, did we loose something else ?

**Review by gares — COMMENTED** [review]
Some comments from the coffee room

File: proofs/arch/arch_extra.v [hunk 5]
```diff
@@ -273,7 +274,7 @@ Variant extended_op :=
 
 Definition extended_op_beq o1 o2 :=
   match o1, o2 with
-  | BaseOp o1, BaseOp o2 => eq_op (T:= prod_eqType _ ceqT_eqType) o1 o2
+  | BaseOp o1, BaseOp o2 => @eq_op (_ * ceqT_eqType)%type o1 o2
```
[the remark this answers is not in the document]
Comment (proux01) [reply proofs/arch/arch_extra.v]: It's now called `s` (for `sort` I guess) for all structures, we should maybe change that but this will require some hacking into hierarchy-builder, not sure it's worth it. Anyway, Enrico's solution is much better here:
```Coq
  | BaseOp o1, BaseOp o2 => o1 == o2 :> _ * ceqT_eqType
```

File: proofs/compiler/x86_instr_decl.v (hunk 1 above)
Comment (proux01) [reply proofs/compiler/x86_instr_decl.v:868]: Yes, `word.word_ringType 0` was a ring structure on `word 0.+1` (which is moral since ring requires `0 != 1` which doesn't hold on empty words), at least this is now much clearer.

File: proofs/lang/memory_model.v (hunk 3 above)
Comment (proux01) [reply proofs/lang/memory_model.v]: `lia` was no longer working because we had for some reason both `GRing_Zmodule__to__GRing_Nmodule (GRing_ComRing__to__GRing_Zmodule (word Uptr))` and `GRing_ComRing__to__GRing_Nmodule (word Uptr)` which are convertible but not syntactically equal, hence not the same variable for lia. Maybe using https://github.com/math-comp/mczify/ would help but not sure since there is a mix of stdlib and mathcomp in those arithmetic expressions.
I pushed a (slightly) better fix for now.

File: proofs/lang/word.v (hunk 4 above)
Comment (proux01) [reply proofs/lang/word.v:2139]: The `apply: Zmult_lt_compat_r` was no longer working (or terribly slow) I didn't really investigate why. There might be a simpler fix.

File: proofs/lang/memory_model.v [hunk 6]
```diff
@@ -722,7 +722,8 @@ Proof.
...
+  by apply: Z.le_lt_trans hlt; apply: (Z.le_sub_nonneg _ _).1.
```
[the remark this answers is not in the document]
Comment (proux01) [reply proofs/lang/memory_model.v]: C.f. above https://github.com/jasmin-lang/jasmin/pull/560#discussion_r1317487810

File: proofs/arch/arch_extra.v (hunk 5 above)
[the remark this answers is not in the document]
Comment (proux01) [reply proofs/arch/arch_extra.v]: We could indeed fix that in MC2, not sure it's worth the effort though.
Your solution here is much better anyway, I kept it.

**Comment by proux01** [comment]
> what is the level of maturity of mathcomp2 ?

A number of major reverse dependencies have already been successfully ported: abel, analysis, comp-dec-modal, coqeal, coquelicot, finmap, fourcolor, gaia, graph-theory, interval, mczify, multinomials, odd-order, real-closed, reglang, tarjan and others work with both MC1 and MC2 (basically devs that don't instantiate any structure).
It is now out since a bit more than three months. There has been some performance issues that we are still improving (particularly with Coq 8.17.0, fixed in 8.17.1). Mathcomp 1 remains maintained and features easily backported will be available there but major new developments now happen on MC2, so no hurry.

> I notice that our code is simplified but except that did you gain other things with this change ?

Not much. There are new structures for semirings and semigroups (useful for instance for bigop with min and max when they don't have a neutral element and hence cannot be equipped with a monoid structure) that you may benefit. But honestly, my main motivation was that Jasmin is now among the main unported dev both in nixpkgs and Coq CI so if we want MC2 to be made the default there, this will have to be ported at some point.

> I notice that we need a much recent version of Coq,

Coq >= 8.16, that's indeed one year more recent than 8.14 currently required by Jasmin.

> did we loose something else ?

MC2 requires hierarchy-builder, itself requiring coq-elpi to build so the dependency chain is a bit longer. But that's more mathcomp developers' problem than yours.

Compilation time and memory use might also be a bit larger: on mathcomp itself and mathcomp Analysis, compilation typically requires 30 to 40% more memory and takes a bit less than 3 times longer, but those are heavy users of elaborate structures, for you use case the impact is probably much less.

**Comment by vbgl** [comment]
Sure. Thanks for your help. I’ve just rebased and launched CI.

**Comment by gares** [comment]
> Compilation time and memory use might also be a bit larger: on mathcomp itself and mathcomp Analysis, compilation typically requires 30 to 40% more memory and takes a bit less than 3 times longer, but those are heavy users of elaborate structures, for you use case the impact is probably much less.

I think we need to look into it in order to know what happens in this specific case. Sometimes the new layout of structures/classes (that is flat) is a bit worse in terms of term comparison performance. The issue popped up a couple of times in MC. On the good side in MC2 we have a very handy locking mechanism which  helps to hide "large terms" behind opaque symbols.

**Comment by proux01** [comment]
Coq 8.17.0 is known to have a dramatic memory issue, could you try with another version?

**Comment by gares** [comment]
I did not spot that. If you are with that combo of coq, coq-elpi and HB the perfs are very bad.

**Comment by proux01** [comment]
Indeed, apparently it's much better.
Let me pin jasmin in Coq's CI and as far as I'm concerned, this is mergable.

**Comment by vbgl** [comment]
> Indeed, apparently it's much better. 

How do you know? What’s a good way to measure time & memory consumption of `coqc`?

**Comment by proux01** [comment]
> > Indeed, apparently it's much better.
> 
> How do you know?

Well, it's working whereas it wasn't but I don't know more.

> What’s a good way to measure time & memory consumption of `coqc`?

When using coq_makefile, doing `make -j1 TIMED=1` works fine.

**Comment by proux01** [comment]
@vbgl with https://github.com/coq/coq/pull/18402 merged, merging the current PR won't break Coq's CI.

**Comment by vbgl** [comment]
What’s the meaning of all these new warnings? Should we pay attention to them or just silence them?

**Comment by proux01** [comment]
Which one?

**Comment by vbgl** [comment]
All of them. Here is one, for instance:

~~~
File "./lang/utils.v", line 77, characters 0-90:
Warning: Ignoring canonical projection to isCountable.pickleK by
Choice_isCountable.pickleK in
choice_isCountable__to__choice_Choice_isCountable: redundant with
choice.choice_isCountable__to__choice_Choice_isCountable
[redundant-canonical-projection,records,default]
~~~

**Comment by proux01** [comment]
> All of them.

I was asking because I no longer see any CI result here.

> Here is one, for instance:
> 
> ```
> File "./lang/utils.v", line 77, characters 0-90:
> Warning: Ignoring canonical projection to isCountable.pickleK by
> Choice_isCountable.pickleK in
> choice_isCountable__to__choice_Choice_isCountable: redundant with
> choice.choice_isCountable__to__choice_Choice_isCountable
> [redundant-canonical-projection,records,default]
> ```

I see, the fact that those warnings are displayed is a bug of HB, they should be ignored, for the record, here is what we have in our _CoqProject for mathcomp:
```
-arg -w -arg -projection-no-head-constant
-arg -w -arg -redundant-canonical-projection
-arg -w -arg -notation-overridden
```

**Comment by vbgl** [comment]
Thanks. I’ve thus disable this warning, rebased, and launched CI.

There are some deprecated notations that should be fixed.

File: proofs/compiler/stack_alloc.v [hunk 7]
```diff
@@ -82,12 +83,11 @@ Proof.
...
 Qed.
 
-Definition region_eqMixin := Equality.Mixin region_axiom.
-Canonical  region_eqType  := Eval hnf in EqType region region_eqMixin.
+HB.instance Definition _ := hasDecEq.Build region region_axiom.
 
 Module CmpR.
 
-  Definition t := [eqType of region].
+  Definition t := Equality.clone _ region.
```
Comment (proux01) [inline proofs/compiler/stack_alloc.v:90]: It's easier to just
```suggestion
  Definition t : eqType := region.
```

Comment (vbgl) [reply proofs/compiler/stack_alloc.v]: Thanks. The deprecation message might be enhanced then.

Do you have suggestions for the remaining explicit uses of `Equality.clone`?

Comment (proux01) [reply proofs/compiler/stack_alloc.v]: > Thanks. The deprecation message might be enhanced then.

Yeah, unfortunately we never came up with a good wording for those deprecation messages.
 
> Do you have suggestions for the remaining explicit uses of `Equality.clone`?

Basically, thanks to reverse coercions, having a type constraint around should be enough.

File: proofs/lang/warray_.v [hunk 8]
```diff
@@ -51,7 +51,7 @@ Module WArray.
   Definition empty (s:positive) : array s :=
     {| arr_data := Mz.empty _ |}.
 
-  Local Notation pointer := [eqType of Z].
+  Local Notation pointer := (Equality.clone _ Z).
```
Comment (proux01) [inline proofs/lang/warray_.v:54]: Just nitpicking, haven't tried but
```suggestion
  Local Notation pointer := Z.
```
may work (or `Z : eqType` otherwise)

Comment (vbgl) [reply proofs/lang/warray_.v]: Thanks a lot!

File: proofs/arch/asm_gen.v [hunk 9]
```diff
@@ -589,13 +589,13 @@ Definition assemble_prog (p : lprog) : cexec asm_prog :=
...
-      ((to_reg  rip == None :> option_eqType ceqT_eqType) && 
-       (to_regx rip == None :> option_eqType ceqT_eqType))  
+      ((to_reg  rip == None :> (option ceqT_eqType : eqType)) && 
+       (to_regx rip == None :> (option ceqT_eqType : eqType)))  
       (E.gen_error true None None (pp_s "Invalid RIP"))
   in
   Let _ :=
     assert
-      (of_ident (lp_rsp p) == Some ad_rsp :> option_eqType ceqT_eqType)
+      (of_ident (lp_rsp p) == Some ad_rsp :> (option ceqT_eqType : eqType))
```
Comment (gares) [inline proofs/arch/asm_gen.v:598]: I suspect `(option ceqT : eqType)` to be sufficient.
I also suspect `of_ident (lp_rsp p) == Some ad_rsp :> option ceqT` to be sufficient, since `==` will impose the same type constraint.

Comment (vbgl) [reply proofs/arch/asm_gen.v:598]: There is no `ceqT`. But indeed, this can be made slightly simpler.

**Comment by vbgl** [comment]
Some timing measurements were done with with `make -j1 TIMED=1` on main (c262bdfa5fb11ae87f036409663608a4fbd6fdfc) and on this branch, on an otherwise idle machine. Here are a few selected highlight.

Total run time: before 11mn, after 17mn.
Total memory (GiO): before 93, after 123.

Before:
arch/arch_decl.vo (real: 1.82, user: 1.73, sys: 0.08, mem: 698964 ko)
lang/expr.vo (real: 2.68, user: 2.56, sys: 0.11, mem: 720996 ko)
lang/warray_.vo (real: 3.57, user: 3.43, sys: 0.14, mem: 653992 ko)
lang/slh_ops.vo (real: 1.14, user: 1.00, sys: 0.13, mem: 644856 ko)

After:
arch/arch_decl.vo (real: 9.99, user: 9.82, sys: 0.17, mem: 1152512 ko)
lang/expr.vo (real: 9.73, user: 9.56, sys: 0.16, mem: 1175444 ko)
lang/warray_.vo (real: 7.27, user: 7.09, sys: 0.17, mem: 948232 ko)
lang/slh_ops.vo (real: 5.50, user: 5.34, sys: 0.16, mem: 933132 ko)

**Comment by proux01** [comment]
With which version of Coq \ coq-elpi \ hierarchy-builder? In any case, that seems typical of what we observe on mathcomp or mathcomp analysis themselves.

**Comment by vbgl** [comment]
Coq 8.18, mathcomp 2.1.0, hb 1.5.0, coq-elpi 1.19.0 (ocaml-elpi 1.17.0)

Good to know it is typical. But it is acceptable?

**Comment by proux01** [comment]
Those are indeed the last released versions. We still have ideas for improvement so it is likely to further improve in the future. For instance, the forthcoming release of HB will come with https://github.com/math-comp/hierarchy-builder/pull/391

**Comment by vbgl** [comment]
I think we’ll wait a bit before moving to mathcomp-2. Currently, the performance regression is to large, and the benefits are not clear.

**Comment by proux01** [comment]
Have you tried with the latest hierarchy-builder 1.7? this may bring interesting performance improvements.

**Comment by spitters** [comment]
What is the status of this? We have updated SSProve to MC2.0 and would also like to update the jasmin part.

**Comment by vbgl** [comment]
This is stuck due to performance regressions that would be induced by the move to MC2.

**Comment by proux01** [comment]
Is the performance regression still unacceptable with hierarchy-builder 1.7?

**Comment by vbgl** [comment]
I’ve spent too much time on this PR already; I may give a try to hb1.7 in a couple of months.

**Comment by gares** [comment]
No problem, this will give us some time to look at the actual regression.

**Comment by spitters** [comment]
> Is the performance regression still unacceptable with hierarchy-builder 1.7?

@cmester0 mentioned that HB 1.7 still has similar timings on his machine :-(

**Comment by vbgl** [comment]
Just rebased after the « less algebra » PR. Ooh, CI failed:

> 11680 Killed

**Comment by gares** [comment]
A short investigation hints that HB.instance is slow. We know that, and the only workaround (for now) is to minimize imports.

With this patch:
```diff
diff --git a/proofs/arch/arch_decl.v b/proofs/arch/arch_decl.v
index f607225e..81f8b430 100644
--- a/proofs/arch/arch_decl.v
+++ b/proofs/arch/arch_decl.v
@@ -1,6 +1,6 @@
 (* -------------------------------------------------------------------- *)
 From HB Require Import structures.
-From mathcomp Require Import all_ssreflect ssralg ssrnum.
+From mathcomp Require Import ssreflect ssrfun ssrbool ssrnat eqtype ssralg.
 From mathcomp Require Import word_ssrZ.
 From Coq Require Import
   Relation_Operators
```
I go from
```
$ time make arch/arch_decl.vo
make -f Makefile.coq arch/arch_decl.vo
make[1]: Entering directory '/home/tassi/COQ/master/_build_ci/jasmin/proofs'
COQDEP VFILES
COQC arch/arch_decl.v
make[1]: Leaving directory '/home/tassi/COQ/master/_build_ci/jasmin/proofs'

real    0m7,645s
user    0m7,219s
sys     0m0,386s
```
to
```
$ time make arch/arch_decl.vo
make -f Makefile.coq arch/arch_decl.vo
make[1]: Entering directory '/home/tassi/COQ/master/_build_ci/jasmin/proofs'
COQDEP VFILES
COQC arch/arch_decl.v
make[1]: Leaving directory '/home/tassi/COQ/master/_build_ci/jasmin/proofs'

real    0m3,863s
user    0m3,543s
sys     0m0,319s
```
since `order.vo` declares a load of structures, and `HB.instance` has a bad complexity on that factor.

**Comment by gares** [comment]
```diff
diff --git a/proofs/lang/warray_.v b/proofs/lang/warray_.v
index a71632d4..9b1b32b8 100644
--- a/proofs/lang/warray_.v
+++ b/proofs/lang/warray_.v
@@ -3,7 +3,7 @@
 (* ** Imports and settings *)
 Require Export ZArith Setoid Morphisms.
 From HB Require Import structures.
-From mathcomp Require Import all_ssreflect ssralg ssrnum.
+From mathcomp Require Import ssreflect ssrfun ssrbool ssrnat eqtype div ssralg.
 From mathcomp Require Import word_ssrZ.
 Require Import xseq.
 Require Export utils array gen_map type word memory_model.
```
From
```
$ time make -C proofs lang/warray_.vo
make: Entering directory '/home/tassi/COQ/master/_build_ci/jasmin/proofs'
make -f Makefile.coq lang/warray_.vo
make[1]: Entering directory '/home/tassi/COQ/master/_build_ci/jasmin/proofs'
COQC lang/warray_.v
make[1]: Leaving directory '/home/tassi/COQ/master/_build_ci/jasmin/proofs'
make: Leaving directory '/home/tassi/COQ/master/_build_ci/jasmin/proofs'

real    0m6,251s
user    0m5,950s
sys     0m0,298s
```
To
```
$ time make -C proofs lang/warray_.vo
make: Entering directory '/home/tassi/COQ/master/_build_ci/jasmin/proofs'
make -f Makefile.coq lang/warray_.vo
make[1]: Entering directory '/home/tassi/COQ/master/_build_ci/jasmin/proofs'
COQDEP VFILES
COQC lang/warray_.v
make[1]: Leaving directory '/home/tassi/COQ/master/_build_ci/jasmin/proofs'
make: Leaving directory '/home/tassi/COQ/master/_build_ci/jasmin/proofs'

real    0m4,796s
user    0m4,493s
sys     0m0,304s
```

**Comment by gares** [comment]
```diff
$ git diff proofs/lang/slh_ops.v
diff --git a/proofs/lang/slh_ops.v b/proofs/lang/slh_ops.v
index 9842e706..97fb5407 100644
--- a/proofs/lang/slh_ops.v
+++ b/proofs/lang/slh_ops.v
@@ -1,6 +1,6 @@
 From HB Require Import structures.
 From mathcomp Require Import
-  all_ssreflect
+  ssreflect ssrfun ssrbool ssrnat div eqtype
   ssralg ssrnum.
 
 Require Import
```
From
```
$ time make -C proofs lang/slh_ops.vo
make: Entering directory '/home/tassi/COQ/master/_build_ci/jasmin/proofs'
make -f Makefile.coq lang/slh_ops.vo
make[1]: Entering directory '/home/tassi/COQ/master/_build_ci/jasmin/proofs'
COQC lang/sem_type.v
COQC lang/slh_ops.v
make[1]: Leaving directory '/home/tassi/COQ/master/_build_ci/jasmin/proofs'
make: Leaving directory '/home/tassi/COQ/master/_build_ci/jasmin/proofs'

real    0m8,422s
user    0m7,923s
sys     0m0,500s
```
To
```
tassi@babum:~/COQ/master/_build_ci/jasmin$ time make -C proofs lang/slh_ops.vo
make: Entering directory '/home/tassi/COQ/master/_build_ci/jasmin/proofs'
make -f Makefile.coq lang/slh_ops.vo
make[1]: Entering directory '/home/tassi/COQ/master/_build_ci/jasmin/proofs'
COQDEP VFILES
COQC lang/slh_ops.v
make[1]: Leaving directory '/home/tassi/COQ/master/_build_ci/jasmin/proofs'
make: Leaving directory '/home/tassi/COQ/master/_build_ci/jasmin/proofs'

real    0m2,970s
user    0m2,673s
sys     0m0,290s
```

**Comment by gares** [comment]
OK, so the pattern is clear and we know what to optimize in HB. It also slows down other works and we are planning a sprint.
We are also considering moving order.v from all_ssreflect to a dedicated package.

I have no idea if the work around I propose here is acceptable and can move this PR forward, or if overall it is still too slow.
@proux01 would you mind integrating these changes in the PR when you have the time?

**Comment by proux01** [comment]
@gares done, here are the figures:
* branch `main`: 10:50, 2604 MB
* this branch: 12:54, 3077 MB

@vbgl is this <= 20% overhead acceptable?

**Comment by vbgl** [comment]
I’d be glad to get your opinion (Pierre, Enrico) on that other PR: https://github.com/jasmin-lang/coqword/pull/30.

Note that `mathcomp-word` uses `order`, therefore whatever package this library belongs to is a dependency of Jasmin.

**Comment by proux01** [comment]
> I’d be glad to get your opinion (Pierre, Enrico) on that other PR: [jasmin-lang/coqword#30](https://github.com/jasmin-lang/coqword/pull/30).

I did also try that but didn't go very far. It cannot harm but seems less useful as it doesn't reduces the imports that much.

> Note that `mathcomp-word` uses `order`, therefore whatever package this library belongs to is a dependency of Jasmin.

Note that there is a difference between requiring and importing. Sure, order will be transitively required but as Enrico said the point is to no import it.

**Comment by eponier** [comment]
I merged PR #801. [The CI](https://gitlab.com/jasmin-lang/jasmin/-/jobs/6897666632) reveals an error with version 2.2.0 of mathcomp that I don't know how to fix.

**Comment by proux01** [comment]
Looks like a Nix issue, you should ask @vbgl I guess.

**Comment by vbgl** [comment]
The fix is to wait for latest mathcomp to be available in CI. This might take a couple of weeks and require to address #803.

**Comment by eponier** [comment]
Sorry, I don't get it, what is the link between the latest mathcomp and this Jasmin bug?

**Comment by vbgl** [comment]
Updating the CI environment might bring in recent binutils.

**Comment by vbgl** [comment]
Rebased and run the measurements again.

Total run time: before 11mn, after 14mn.
Total memory (GiO): before 91, after 117.

I can no longer find any file with a dramatic increase in cost.

**Comment by eponier** [comment]
Ouch:
- job coq-program: 25mn instead of 5mn
- job coq-proof: 30mn instead of 10mn
Why is this so different from Pierre's numbers? I noticed you rolled back to mathcomp 2.1. Does this mean that we really need 2.2?

**Comment by vbgl** [comment]
I think you are comparing the time to build math-comp 2 vs the time to download math-comp 1. This does not look fair.

**Comment by eponier** [comment]
@vbgl Now that https://github.com/jasmin-lang/jasmin/pull/822 is merged, can you rebase and select mathcomp 2.2 as a dependency? Or did I misunderstand once more how nix and our CI is working?

**Comment by vbgl** [comment]
I will, but there is no hurry.

**Comment by vbgl** [comment]
Rebased on top of recent `main` branch, updated to use mathcomp 2.2.0.

**Comment by eponier** [comment]
coq-master fails for whatever reason, but the other results are good.
- job coq-program: 7mn instead of 5mn
- job coq-proof: 10mn instead of 10mn

**Comment by gares** [comment]
Note that this very week we are having a sprint to optimize HB.instance, see https://coq.zulipchat.com/#narrow/stream/237868-Hierarchy-Builder-devs-.26-users/topic/Hierarchy.20Builder.20optimization.20sprint

I guess we will make a release of HB soon, although I don't know how much you will gain now that the imports are minimized. For sure you won't lose anything ;-)

**Review by vbgl — APPROVED** [review]

**Comment by eponier** [comment]
> @vbgl please ping me whenever you plan to merge this so that we can avoid breaking Coq's CI too much.

@proux01 About to merge.

**Comment by eponier** [comment]
Thanks everyone involved for your time on this PR!
